### ISSUE TITLE
[Design flaw] write to gpfdist external table: unexpected error HTTP 500 

### ISSUE THREAD

**Issue by kmjs233** [greenplum-db/gpdb#14358]
## Bug Report
[gpfdist] write to external table: unexpected error HTTP 500 

### Greenplum version or build
6.21

### OS version and uname -a
el7.9

### Expected behavior
With heavy load, we expect slowdown, not random errors.

### Actual behavior
Http 500 error observed in production when unloading data with gpfdist/writable external tables.
Write/Post to gpfdist may randomly error out within a few seconds due to retry mechanism in url_curl.c
gpfdist -V log: 
* [WARN] handle post request receive errno: 104, msg: Connection reset by peer
* [WARN] HTTP ERROR: {segment IP} - 500 internal error
* [WARN] gpfdist_send faild - the connection was terminated by the client (32: Broken pipe)
* failed to shutdown socket, errno 107, msg: Transport endpoint is not connected

segment log:
* response (56 - Failure when receiving data from the peer)
* failed to send request to gpfdist
* will retry after 1 seconds
* http response code 500 from gpfdist
* Process interruptfor 'query cancel pending'
* 25M01, canceling MPP operation

By design, gpfdist handles sequetial data by processing requests one by one. Imagine we are reading from a data stream, it's impossible rewind and read old data, unless we save all data. When writing to a data stream, it's impossible to revoke any written data and write once more. 
By reading url_curl.c, and with a few tests, I found out that db segment retries post request if response is not received with in 1/2/4/8/... seconds. Such a retry leads to failure: 
Retry is not an option. A retry resets previous HTTP conversation, no matter queued or being processed in gpfdist. When gpfdist tries to read post data from previous conversation, it will get RESET, and close the session. Any subsequent request to that session will receive HTTP 500 error, the cancel the query. The db segment where retry happens will not receive respone of the original conversation, but a response for the retry. 

By now I saw such errors happen in about 1 minute, not 2 seconds, which I couldn't figure out why. Maybe some other mechanism that delay the timeout check.
What I do know:
* existing GUC gpfdist_retry_timeout doesn't help
* hard-coded MAX_TRY_WAIT_TIME 64 doesn't help
Both are about total time of all tries, not grace time of a single try.

### Step to reproduce the behavior
Basically it happens more easily in big cluster, with heavy load/slow network. When gpfdist/network is very busy and needs plenty of time to respond to all segments.
What may help: 
* large writable_external_table_bufsize, max 128MB
* large number of db segments
* write to multiple external tables in parallel, with the same gpfdist process/port

### [CONCLUSION] 
By the nature of gpfdist, it's not eligible for retry. Retry is wrong design. It does no good, whatever the grace time is.
A query should fail if it doesn't get response from gpfdist within grace time. Hard-coded 1 second causes highly unpredictable errors. It's ideal if only we have a GUC.

**Comment by kmjs233** [comment]
I suspect curl timeout was reset when 
get_gpfdist_status(curl) was called (even if it's designed for readable external tables), probably why neither MAX_TRY_WAIT_TIME nor gpfdist_retry_timeout works:

	if (CURLE_OK != (e = curl_easy_setopt(status_handle, CURLOPT_TIMEOUT, 60L)))

**Comment by kmjs233** [comment]
To reproduce:
make a name pipe and write to / read from it

**Comment by RuiWang-vmw** [comment]
A Jira story has been created to track this issue: https://jira.eng.vmware.com/browse/GPDF-55

**Comment by Hlinbit** [comment]
https://github.com/greenplum-db/gpdb/pull/14443

This PR provides a switch to turn on/off the retry mechanism for writing into the writable external table. By executing the command 'set gpfdist_retry_timeout = 0;', you can turn off the retry. As for whether to cancel the retry, we need more discussion to determine.

**Comment by huiliang-liu** [comment]
once the segments got error 56 - Failure when receiving data from the peer, server should already close/reset the connection. you can capture the tcpdump log to check it. Anyway, the HTTP connection is end, there is no extra action segment side can do. we can't ignore the error and expect gpfdist can receive the correct data.

**Comment by kainwen** [comment]
@huiliang-liu @Hlinbit Several questions, thanks!

* is this a problem?
* does it exist on main branch also?
* will https://github.com/greenplum-db/gpdb/pull/14443 fix this?

**Comment by kmjs233** [comment]
> @huiliang-liu @Hlinbit Several questions, thanks!
> 
> * is this a problem?
> * does it exist on main branch also?
> * will [(WIP)Add the switch for writable external table transfer retry. #14443](https://github.com/greenplum-db/gpdb/pull/14443) fix this?

From what I understand:
* GUC gpfdist_retry_timeout makes no sense because retry is not an option. #14443 is helpful, but might not be a fix. It is to ignore above GUC, and possibly reduces the chance of failures. 
* To make such failures predictable or under control, we need insights on where retry can happen, besides in above scene.

**Comment by huiliang-liu** [comment]
This issue is resolved by PR #15722 which makes sure retry is effective.
It has been verified in the environment as you mentioned @kmjs233 :
- large writable_external_table_bufsize, max 128MB-
- large number of db segments

**Comment by huiliang-liu** [comment]
This issue is fixed in GPDB 6.24.5
